# backtest-case: CXX-fused_topk_with_score_function
# kind: cxx
# base: 42b840051647eef89761a16dfdff87e82bb253ab
# derived: diff(BASE:transformer_engine/common/fused_router/fused_topk_with_score_function.cu, PRE:transformer_engine/common/fused_router/fused_topk_with_score_function.cu)
--- a/transformer_engine/common/fused_router/fused_topk_with_score_function.cu
+++ b/transformer_engine/common/fused_router/fused_topk_with_score_function.cu
@@ -1,4 +1,6 @@
 /*************************************************************************
+ * This file was modified for portability to AMDGPU
+ * Copyright (c) 2025-2026, Advanced Micro Devices, Inc. All rights reserved.
  * Copyright (c) 2022-2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
  *
  * See LICENSE for license information.
@@ -155,7 +157,12 @@
         __syncwarp();
         // Compute the group score
         if (lane_id == 0) {
+//TODO: release after /opt/rocm/include/hip/amd_detail/amd_hip_bfloat16.h remove explict constructor restriction
+#ifdef __HIP_PLATFORM_AMD__
+          CompType tmp(0.0f);
+#else
           CompType tmp = 0.0;
+#endif
           for (int j = 0; j < topk / group_topk; j++) {
             tmp = tmp + topk_scores[j];
           }
@@ -401,7 +408,11 @@
       }
       // Warp reduce the sum
       for (int s = 16; s > 0; s /= 2) {
+#ifdef __HIP_PLATFORM_AMD__
+        local_sum_Output_x_Grad += __shfl_xor(local_sum_Output_x_Grad, s, kThreadsPerWarp);
+#else
         local_sum_Output_x_Grad += __shfl_xor_sync(0xffffffff, local_sum_Output_x_Grad, s);
+#endif
       }
       CompType sum_Output_x_Grad = local_sum_Output_x_Grad;
       // In-place update
